File: tests/test_dispatch_ledger.py
"""Tests for dispatch_ledger.py."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

import unittest
from harness_core.dispatch.dispatch_decision import BudgetReservation, DispatchDecision
from harness_core.dispatch.dispatch_ledger import DispatchBundle, DispatchLedger
from harness_core.dispatch.evaluation_stub import EvaluationCheck, EvaluationResult
from harness_core.dispatch.executor_adapter import ExecutionResult
from harness_core.dispatch.task_analyzer import RuleBasedTaskAnalyzer, TaskAnalysis


class DispatchLedgerTests(unittest.TestCase):
    def setUp(self):
        self.ledger = DispatchLedger()

    def test_create_record(self):
        r = self.ledger.create_record("disp-001", "test request", "a-001", "dec-001")
        self.assertEqual(r.dispatch_id, "disp-001")
        self.assertEqual(r.final_status, "dispatched")

    def test_get_record(self):
        self.ledger.create_record("disp-001", "req", "a-001", "dec-001")
        r = self.ledger.get_record("disp-001")
        self.assertIsNotNone(r)
        self.assertEqual(r.dispatch_id, "disp-001")

    def test_get_nonexistent_record(self):
        r = self.ledger.get_record("nonexistent")
        self.assertIsNone(r)

    def test_list_records(self):
        self.ledger.create_record("disp-001", "req1", "a-001", "dec-001")
        self.ledger.create_record("disp-002", "req2", "a-002", "dec-002")
        records = self.ledger.list_records()
        self.assertEqual(len(records), 2)

    def test_update_record(self):
        r = self.ledger.create_record("disp-001", "req", "a-001", "dec-001")
        updated = self.ledger.update_record(r, final_status="completed", execution_result_id="exec-001")
        self.assertEqual(updated.final_status, "completed")
        self.assertEqual(updated.execution_result_id, "exec-001")
        self.assertNotEqual(updated.updated_at, r.created_at)

    def test_update_persists(self):
        r = self.ledger.create_record("disp-001", "req", "a-001", "dec-001")
        self.ledger.update_record(r, final_status="completed")
        stored = self.ledger.get_record("disp-001")
        self.assertEqual(stored.final_status, "completed")

    def test_replay(self):
        self.ledger.create_record("disp-001", "req", "a-001", "dec-001")
        r = self.ledger.replay("disp-001")
        self.assertIsNotNone(r)
        self.assertEqual(r.dispatch_id, "disp-001")

    def test_to_dict(self):
        r = self.ledger.create_record("disp-001", "req", "a-001", "dec-001")
        d = r.to_dict()
        self.assertIn("dispatch_id", d)
        self.assertIn("final_status", d)
        self.assertIn("schema_version", d)

    def test_create_with_budget_reservation(self):
        r = self.ledger.create_record("disp-001", "req", "a-001", "dec-001",
                                       budget_reservation_id="bres-001")
        self.assertEqual(r.budget_reservation_id, "bres-001")

    def test_update_with_usage_ledger_row(self):
        r = self.ledger.create_record("disp-001", "req", "a-001", "dec-001")
        updated = self.ledger.update_record(r, usage_ledger_row_id="ulr-001")
        self.assertEqual(updated.usage_ledger_row_id, "ulr-001")


def _make_analysis():
    return RuleBasedTaskAnalyzer().analyze("Review auth.py for security issues")


def _make_decision():
    return DispatchDecision(
        decision_id="dec-001",
        analysis_id="a-001",
        analysis_snapshot={},
        selected_tier="balanced_worker",
        fallback_tier="cheap_executor",
        routing_reason="test",
        quality_requirement="standard",
        expected_quality_band="medium",
        confidence=0.8,
        confidence_label="high",
        budget_reservation=BudgetReservation(
            reservation_id="r-001", decision_id="dec-001", currency="token",
            pre_budget=5000, reserved_input_tokens=3000, reserved_output_tokens=2000,
            reserved_total_tokens=5000, reserved_cost=0.05, status="reserved",
            created_at="2026-01-01T00:00:00Z", updated_at="2026-01-01T00:00:00Z",
        ),
        execution_policy={"executor_type": "noop", "execution_allowed": True, "requires_human_review": False, "max_retries": 0},
        decision_status="decided",
        created_at="2026-01-01T00:00:00Z",
    )


def _make_execution_result():
    return ExecutionResult(
        result_id="exec-001",
        dispatch_id="disp-001",
        decision_id="dec-001",
        executor_type="noop",
        status="completed",
        output="test output",
        created_at="2026-01-01T00:00:00Z",
    )


def _make_evaluation_result():
    return EvaluationResult(
        evaluation_id="eval-001",
        dispatch_id="disp-001",
        decision_id="dec-001",
        execution_result_id="exec-001",
        status="pass",
        checks=(
            EvaluationCheck(check_id="c1", name="schema_ok", status="pass", reason="ok"),
        ),
        created_at="2026-01-01T00:00:00Z",
    )


class DispatchBundleTests(unittest.TestCase):
    def setUp(self):
        self.ledger = DispatchLedger()
        self.record = self.ledger.create_record("disp-001", "req", "a-001", "dec-001")

    def test_store_bundle(self):
        bundle = self.ledger.store_bundle(
            self.record, _make_analysis(), _make_decision(),
            _make_execution_result(), _make_evaluation_result(),
        )
        self.assertIsInstance(bundle, DispatchBundle)
        self.assertEqual(bundle.record.dispatch_id, "disp-001")

    def test_get_bundle(self):
        self.ledger.store_bundle(
            self.record, _make_analysis(), _make_decision(),
            _make_execution_result(), _make_evaluation_result(),
        )
        bundle = self.ledger.get_bundle("disp-001")
        self.assertIsNotNone(bundle)
        self.assertEqual(bundle.record.dispatch_id, "disp-001")

    def test_get_nonexistent_bundle(self):
        self.assertIsNone(self.ledger.get_bundle("missing"))

    def test_list_bundles(self):
        self.ledger.store_bundle(
            self.record, _make_analysis(), _make_decision(),
            _make_execution_result(), _make_evaluation_result(),
        )
        bundles = self.ledger.list_bundles()
        self.assertEqual(len(bundles), 1)

    def test_replay_returns_bundle_when_stored(self):
        self.ledger.store_bundle(
            self.record, _make_analysis(), _make_decision(),
            _make_execution_result(), _make_evaluation_result(),
        )
        result = self.ledger.replay("disp-001")
        self.assertIsInstance(result, DispatchBundle)

    def test_replay_returns_record_when_no_bundle(self):
        result = self.ledger.replay("disp-001")
        self.assertIsInstance(result, type(self.record))
        self.assertNotIsInstance(result, DispatchBundle)

    def test_bundle_to_dict(self):
        bundle = self.ledger.store_bundle(
            self.record, _make_analysis(), _make_decision(),
            _make_execution_result(), _make_evaluation_result(),
        )
        d = bundle.to_dict()
        self.assertIn("record", d)
        self.assertIn("analysis", d)
        self.assertIn("decision", d)
        self.assertIn("execution_result", d)
        self.assertIn("evaluation_result", d)

File: engine/tests/test_dispatch_ledger.rs
use engine::dispatch_decision::{BudgetReservation, DispatchDecision};
use engine::dispatch_ledger::DispatchLedger;
use engine::evaluation_stub::EvaluationResult;
use engine::executor_adapter::ExecutionResult;
use engine::runtime::FixtureRuntime;
use engine::task_analyzer::{RuleBasedTaskAnalyzer, TaskAnalysis};

fn make_analysis() -> TaskAnalysis {
    RuleBasedTaskAnalyzer::new().analyze("Review auth.py for security issues", "test_fixture")
}

fn make_decision() -> DispatchDecision {
    DispatchDecision {
        decision_id: "dec-001".to_string(),
        analysis_id: "a-001".to_string(),
        analysis_snapshot: serde_json::json!({}),
        selected_tier: "balanced_worker".to_string(),
        fallback_tier: "cheap_executor".to_string(),
        routing_reason: "test".to_string(),
        quality_requirement: "standard".to_string(),
        expected_quality_band: "medium".to_string(),
        confidence: 0.8,
        confidence_label: "high".to_string(),
        budget_reservation: BudgetReservation::default(),
        execution_policy: serde_json::json!({}),
        decision_status: "decided".to_string(),
        created_at: "2026-01-01T00:00:00Z".to_string(),
        ..Default::default()
    }
}

fn make_execution_result() -> ExecutionResult {
    ExecutionResult {
        schema_version: "execution_result.v1".to_string(),
        result_id: "exec-001".to_string(),
        dispatch_id: "disp-001".to_string(),
        decision_id: "dec-001".to_string(),
        executor_type: "noop".to_string(),
        status: "completed".to_string(),
        output: Some("test output".to_string()),
        prompt_pack: None,
        input_tokens: None,
        output_tokens: None,
        estimated_cost: None,
        latency_ms: None,
        error_domain: None,
        error_message: None,
        provider_request_id: None,
        attempt_number: None,
        finish_reason: None,
        usage_source: None,
        created_at: "2026-01-01T00:00:00Z".to_string(),
    }
}

fn make_evaluation_result() -> EvaluationResult {
    EvaluationResult {
        schema_version: "evaluation_result.v1".to_string(),
        evaluation_id: "eval-001".to_string(),
        dispatch_id: "disp-001".to_string(),
        decision_id: "dec-001".to_string(),
        execution_result_id: "exec-001".to_string(),
        status: "pass".to_string(),
        checks: vec![],
        quality_score: None,
        requires_retry: false,
        retry_reason: None,
        created_at: "2026-01-01T00:00:00Z".to_string(),
    }
}

// DispatchRecord tests
#[test]
fn test_create_record() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let r = ledger.create_record("disp-001", "test request", "a-001", "dec-001", None, &rt);
    assert_eq!(r.dispatch_id, "disp-001");
    assert_eq!(r.final_status, "dispatched");
}

#[test]
fn test_create_with_budget_reservation() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let r = ledger.create_record(
        "disp-001",
        "req",
        "a-001",
        "dec-001",
        Some("bres-001".to_string()),
        &rt,
    );
    assert_eq!(r.budget_reservation_id, Some("bres-001".to_string()));
}

#[test]
fn test_update_record() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let r = ledger.create_record("disp-001", "req", "a-001", "dec-001", None, &rt);
    let updated = ledger.update_record(
        r,
        "completed",
        Some("exec-001".to_string()),
        None,
        None,
        &rt,
    );
    assert_eq!(updated.final_status, "completed");
    assert_eq!(updated.execution_result_id, Some("exec-001".to_string()));
}

#[test]
fn test_update_with_usage_ledger_row() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let r = ledger.create_record("disp-001", "req", "a-001", "dec-001", None, &rt);
    let updated =
        ledger.update_record(r, "completed", None, None, Some("ulr-001".to_string()), &rt);
    assert_eq!(updated.usage_ledger_row_id, Some("ulr-001".to_string()));
}

#[test]
fn test_record_to_value() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let r = ledger.create_record("disp-001", "req", "a-001", "dec-001", None, &rt);
    let v = r.to_value();
    assert!(v.get("dispatch_id").is_some());
    assert!(v.get("final_status").is_some());
    assert!(v.get("schema_version").is_some());
}

// DispatchBundle tests
#[test]
fn test_store_bundle() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let record = ledger.create_record("disp-001", "req", "a-001", "dec-001", None, &rt);
    let bundle = ledger.store_bundle(
        record,
        make_analysis(),
        make_decision(),
        make_execution_result(),
        make_evaluation_result(),
    );
    let v = bundle.to_value();
    assert!(v.get("record").is_some());
    assert!(v.get("analysis").is_some());
    assert!(v.get("decision").is_some());
    assert!(v.get("execution_result").is_some());
    assert!(v.get("evaluation_result").is_some());
}

#[test]
fn test_bundle_record_has_dispatch_id() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let record = ledger.create_record("disp-001", "req", "a-001", "dec-001", None, &rt);
    let bundle = ledger.store_bundle(
        record,
        make_analysis(),
        make_decision(),
        make_execution_result(),
        make_evaluation_result(),
    );
    assert_eq!(bundle.record["dispatch_id"], "disp-001");
}

#[test]
fn test_update_preserves_existing_ids() {
    let ledger = DispatchLedger::new();
    let rt = FixtureRuntime::new();
    let r = ledger.create_record(
        "disp-001",
        "req",
        "a-001",
        "dec-001",
        Some("bres-001".to_string()),
        &rt,
    );
    let updated = ledger.update_record(
        r,
        "completed",
        Some("exec-001".to_string()),
        Some("eval-001".to_string()),
        None,
        &rt,
    );
    assert_eq!(updated.budget_reservation_id, Some("bres-001".to_string()));
    assert_eq!(updated.execution_result_id, Some("exec-001".to_string()));
    assert_eq!(updated.evaluation_result_id, Some("eval-001".to_string()));
}
